crates/cgp-field-macro/src/lib.rs

```diff
@@ -1,29 +1,194 @@
+/*!
+   Procedural macros for field manipulation in the CGP framework.
+
+   This crate provides macros for working with fields in algebraic data types,
+   enabling type-safe field access, product and sum types, and symbol handling.
+
+   # Macros
+
+   ## Field Access
+
+   * [`macro@HasField`] - Derive macro for implementing field access traits.
+     Enables type-safe field access and manipulation.
+
+   ## Type Construction
+
+   * [`macro@Product`] - Macro for defining product types (structs).
+   * [`macro@Sum`] - Macro for defining sum types (enums).
+   * [`macro@product`] - Macro for constructing product type expressions.
+
+   ## Symbol Handling
+
+   * [`macro@symbol`] - Macro for creating type-safe symbol references.
+
+   # Examples
+
+   ## Field Access
+
+   ```rust,ignore
+   use cgp_field_macro::HasField;
+
+   #[derive(HasField)]
+   struct Person {
+       name: String,
+       age: u32,
+   }
+   ```
+
+   ## Product Types
+
+   ```rust,ignore
+   use cgp_field_macro::Product;
+
+   Product! {
+       struct Point {
+           x: f64,
+           y: f64,
+       }
+   }
+   ```
+
+   ## Sum Types
+
+   ```rust,ignore
+   use cgp_field_macro::Sum;
+
+   Sum! {
+       enum Shape {
+           Circle(f64),
+           Rectangle(f64, f64),
+       }
+   }
+   ```
+
+   ## Symbol Creation
+
+   ```rust,ignore
+   use cgp_field_macro::symbol;
+
+   let name_symbol = symbol!(name);
+   ```
+*/
+
 extern crate proc_macro;
 
 use proc_macro::TokenStream;
 
+/// Derive macro for implementing field access traits.
+///
+/// This macro generates implementations for field access traits, enabling
+/// type-safe access to struct fields. It analyzes the struct's fields and
+/// generates the necessary trait implementations for each field.
+///
+/// # Examples
+///
+/// ```rust,ignore
+/// #[derive(HasField)]
+/// struct Person {
+///     name: String,
+///     age: u32,
+/// }
+///
+/// // Generated code enables field access like:
+/// let person = Person { name: "Alice".into(), age: 30 };
+/// let name = person.get_field(field!(name));
+/// ```
 #[proc_macro_derive(HasField)]
 pub fn derive_fields(item: TokenStream) -> TokenStream {
     cgp_field_macro_lib::derive_fields(item.into()).into()
 }
 
+/// Macro for creating type-safe symbol references.
+///
+/// This macro creates symbol types that can be used for type-safe field
+/// access and manipulation. Symbols are typically used in conjunction with
+/// field access traits.
+///
+/// # Examples
+///
+/// ```rust,ignore
+/// let name_symbol = symbol!(name);
+/// let age_symbol = symbol!(age);
```

Comment:
The `symbol!` macro is a way to lift string values to the type level. So something like `symbol!("foo")` is a *type* that uniquely represents the string `"foo"`.